src/IceRpc.Slice.Generators/Internal/Parser.cs

```diff
@@ -0,0 +1,202 @@
+    {
+        ImmutableArray<AttributeData> attributes = symbol.GetAttributes();
+        foreach (AttributeData attribute in attributes)
+        {
+            if (SymbolEqualityComparer.Default.Equals(attribute.AttributeClass, attributeSymbol))
+            {
+                return attribute;
+            }
+        }
+        return null;
+    }
+
+    private string GetFullName(ISymbol symbol)
+    {
+        if (symbol is INamespaceSymbol namespaceSymbol && namespaceSymbol.IsGlobalNamespace)
+        {
+            return "";
+        }
+        else
+        {
+            string containingSymbolName = GetFullName(symbol.ContainingSymbol);
+            return containingSymbolName.Length == 0 ? symbol.Name : $"{containingSymbolName}.{symbol.Name}";
+        }
+    }
+
+    private IReadOnlyList<ServiceMethod> GetServiceMethods(ImmutableArray<INamedTypeSymbol> allInterfaces)
+    {
+        Debug.Assert(_operationAttribute is not null);
+        var allServiceMethods = new List<ServiceMethod>();
+        foreach (INamedTypeSymbol interfaceSymbol in allInterfaces)
+        {
+            allServiceMethods.AddRange(GetServiceMethods(interfaceSymbol));
+        }
+        return allServiceMethods;
+    }
+
+    private IReadOnlyList<ServiceMethod> GetServiceMethods(INamedTypeSymbol interfaceSymbol)
+    {
+        var serviceMethods = new List<ServiceMethod>();
+        foreach (IMethodSymbol method in interfaceSymbol.GetMembers().OfType<IMethodSymbol>())
+        {
+            if (GetAttribute(method, _operationAttribute!) is not AttributeData attribute)
+            {
+                continue;
+            }
+
+            foreach (TypedConstant typedConstant in attribute.ConstructorArguments)
+            {
+                if (typedConstant.Kind == TypedConstantKind.Error)
+                {
+                    // if a compilation error was found, no need to keep evaluating other args
+                    return serviceMethods;
+                }
+            }
+
+            ImmutableArray<TypedConstant> items = attribute.ConstructorArguments;
+            Debug.Assert(
+                items.Length == 1,
+                "Unexpected number of arguments in attribute constructor.");
+            string operationName = (string)items[0].Value!;
+            serviceMethods.Add(new ServiceMethod
+            {
+                OperationName = operationName,
+                DispatchMethodName = GetFullName(method)
+            });
+        }
+        return serviceMethods;
+    }
+
+    private bool HasServiceAttribute(INamedTypeSymbol classSymbol)
+    {
+        foreach (AttributeData attribute in classSymbol.GetAttributes())
+        {
+            if (SymbolEqualityComparer.Default.Equals(attribute.AttributeClass, _serviceAttribute))
+            {
+                return true;
+            }
+        }
+
+        if (classSymbol.BaseType is INamedTypeSymbol baseType)
```

Comment:
I don't understand why we do that.

We go out of our way to mark the SliceService attribute Inherited = false, and yet here we implement the equivalent of Inherited = true behavior here.